### PR TITLE
Introduce Keychain API

### LINKED ISSUE #343: Should we allow empty path?
Currently, when an empty path is provided trezor still returns a public key / address. Is that desired or should it return an error?

### PR DESCRIPTION
**Pull request by jpochyla**
Introduces a new way for obtaining keys derived from the master secret. Deriving itself is now synchronous, and namespaced to a set of key-spaces, consisting of curve name and namespacing path.

Fixes #343

### CODE REVIEW

**Comment by jpochyla** [comment]
I'm happy with the code now, requesting review.

File: src/apps/wallet/sign_tx/__init__.py [hunk 1]
```diff
@@ -19,7 +19,8 @@
     coin_name = msg.coin_name or "Bitcoin"
     coin = coins.by_name(coin_name)
     # TODO: rework this so we don't have to pass root to signing.sign_tx
-    root = await seed.derive_node(ctx, [], curve_name=coin.curve_name)
+    keychain = await seed.get_keychain(ctx)
+    root = keychain.derive([], coin.curve_name)
```
[the remark this answers is not in the document]
Comment (jpochyla) [reply src/apps/wallet/sign_tx/__init__.py]: Yes we can!

**Comment by jpochyla** [comment]
Let's include namespace values for all apps in the PR as well.

File: src/apps/ethereum/__init__.py [hunk 2]
```diff
@@ -1,9 +1,12 @@
...
 from trezor.messages import MessageType
 
+from apps.common import HARDENED
+
 
 def boot():
-    wire.add(MessageType.EthereumGetAddress, __name__, "get_address")
-    wire.add(MessageType.EthereumSignTx, __name__, "sign_tx")
-    wire.add(MessageType.EthereumSignMessage, __name__, "sign_message")
+    ns = [["secp256k1", HARDENED | 44, HARDENED | 60]]
```
Comment (tsusanka) [inline src/apps/ethereum/__init__.py]: Could we write `44 | HARDENED` instead (I think it's more readable) or is that black's work?

**Pull request merged**